march_hardware_builder/src/hardware_builder.cpp

```diff
@@ -87,63 +89,71 @@ march::Joint HardwareBuilder::createJoint(const YAML::Node& joint_config, const
   ROS_DEBUG("Starting creation of joint %s", joint_name.c_str());
   HardwareBuilder::validateRequiredKeysExist(joint_config, HardwareBuilder::JOINT_REQUIRED_KEYS, "joint");
 
-  march::ActuationMode mode;
-  if (joint_config["actuationMode"])
-  {
-    mode = march::ActuationMode(joint_config["actuationMode"].as<std::string>());
-  }
-
-  march::IMotionCube imc = HardwareBuilder::createIMotionCube(joint_config["imotioncube"], mode, urdf_joint);
-
-  march::Joint joint(imc);
-  joint.setName(joint_name);
-
-  bool allowActuation = joint_config["allowActuation"].as<bool>();
-  joint.setAllowActuation(allowActuation);
-
+  auto net_number = -1;
   if (joint_config["netNumber"])
   {
-    joint.setNetNumber(joint_config["netNumber"].as<int>());
+    net_number = joint_config["netNumber"].as<int>();
   }
   else
   {
     ROS_WARN("Joint %s does not have a netNumber", joint_name.c_str());
-    joint.setNetNumber(-1);
   }
 
-  if (joint_config["temperatureges"])
+  const auto allow_actuation = joint_config["allowActuation"].as<bool>();
+
+  march::ActuationMode mode;
+  if (joint_config["actuationMode"])
   {
-    march::TemperatureGES ges = HardwareBuilder::createTemperatureGES(joint_config["temperatureges"]);
-    joint.setTemperatureGes(ges);
+    mode = march::ActuationMode(joint_config["actuationMode"].as<std::string>());
   }
-  else
+
+  auto imc = HardwareBuilder::createIMotionCube(joint_config["imotioncube"], mode, urdf_joint);
+  if (!imc)
+  {
+    ROS_WARN("Joint %s does not have a configuration for an IMotionCube", joint_name.c_str());
+  }
+
+  auto ges = HardwareBuilder::createTemperatureGES(joint_config["temperatureges"]);
+  if (!ges)
   {
     ROS_WARN("Joint %s does not have a configuration for a TemperatureGes", joint_name.c_str());
   }
-  return joint;
+  return { joint_name, net_number, allow_actuation, std::move(imc), std::move(ges) };
```

Comment:
Since the `MarchRobot` does not accept pointers to `Joint`s. When I think about it this also makes the joints not injectable with mocks. This is probably because the `MarchRobot` does not have any tests. I will add it to `MarchRobot`

Reply:
Ah I see. This might not be necessary since Joint currently does not directly use any hardware, which should probably be kept this way in the future. It is enough to inject Joint with mock IMCs and mock Encoders

Reply:
Okay I agree